src/GovUk.Education.ExploreEducationStatistics.Notifier.Tests/NotifierFunctionsIntegrationTest.cs

```diff
@@ -1,17 +1,98 @@
+using GovUk.Education.ExploreEducationStatistics.Notifier.Services.Interfaces;
 using Microsoft.Azure.Cosmos.Table;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
+using Moq;
 using Testcontainers.Azurite;
 using Xunit;
 
 namespace GovUk.Education.ExploreEducationStatistics.Notifier.Tests;
 
 public abstract class NotifierFunctionsIntegrationTest
-    (FunctionsIntegrationTestFixture fixture) : FunctionsIntegrationTest<NotifierFunctionsIntegrationTestFixture>(fixture);
+    (NotifierFunctionsIntegrationTestFixture fixture) : FunctionsIntegrationTest<NotifierFunctionsIntegrationTestFixture>(fixture), IAsyncLifetime
+{
+    public Task InitializeAsync()
+    {
+        return Task.CompletedTask;
+    }
+
+    public Task DisposeAsync()
+    {
+        MockUtils.VerifyAllMocks(fixture._emailService);
+        fixture._emailService.Reset();
+
+        return ClearAzureDataTableTestData(TableStorageConnectionString());
+    }
+
+    public string TableStorageConnectionString()
+    {
+        return fixture.TableStorageConnectionString();
+    }
+
+    public AppSettingsOptions GetAppSettingsOptions()
+    {
+        return GetRequiredService<IOptions<AppSettingsOptions>>().Value;
+    }
+
+    public GovUkNotifyOptions GetGovUkNotifyOptions()
+    {
+        return GetRequiredService<IOptions<GovUkNotifyOptions>>().Value;
+    }
+
+    public async Task AddTestSubscription(string tableName, SubscriptionEntity subscription)
+    {
+        var storageAccount = CloudStorageAccount.Parse(TableStorageConnectionString());
+        var tableClient = storageAccount.CreateCloudTableClient();
+        var table = tableClient.GetTableReference(tableName);
+        await table.CreateIfNotExistsAsync();
+
+        await table.ExecuteAsync(TableOperation.InsertOrReplace(subscription));
+    }
+
+    public async Task CreateApiSubscription(ApiSubscription subscription)
+    {
+        var appSettingsOptions = GetAppSettingsOptions();
+
+        var dataTableStorageService = new DataTableStorageService(TableStorageConnectionString());
+
+        await dataTableStorageService.CreateEntityAsync(
+            tableName: appSettingsOptions.ApiSubscriptionsTableName,
+            entity: subscription);
+    }
+
+    public async Task<ApiSubscription?> GetApiSubscriptionIfExists(
+        Guid dataSetId,
+        string email,
+        IEnumerable<string>? select = null)
+    {
+        var appSettingsOptions = GetRequiredService<IOptions<AppSettingsOptions>>();
+
+        var dataTableStorageService = new DataTableStorageService(appSettingsOptions.Value.TableStorageConnectionString);
+
+        return await dataTableStorageService.GetEntityIfExistsAsync<ApiSubscription>(
+            tableName: appSettingsOptions.Value.ApiSubscriptionsTableName,
+            partitionKey: dataSetId.ToString(),
+            rowKey: email,
+            select: select);
+    }
```

Comment:
Could these methods be moved to the test classes that they are used by? E.g. `AddTestSubscription` is only used by `PublicationSubscriptionManagerTests`, but currently you have to navigate to this class to find out what the implementation is.

Reply:
Possibly, however I can see these having the potential to be used elsewhere. For example, when I write the tests for the function that will send notifications - which may be in a different class to the other triggers. I think we do a similar thing in the Processor test project which I think is kinda nice. Somewhat undiscoverable, but the reusability makes up for it in my opinion.

What do you think?

Reply:
I think it's fine to leave these in then, for potential reusability like you say.